### DISCUSSION TITLE
Help with Rascal syntax support that suddenly dropped off

### DISCUSSION THREAD

**Discussion by jurgenvinju** [github-linguist/linguist#6209] [Q&A]
We've had the benefits and pleasure of Rascal syntax highlighting in our projects for a long time with linguist; with the `.rsc` extension.

It does not work anymore for a couple of weeks. The language is detected on the project level and the statistics seem in order, but the GitHub viewers do not offer syntax highlighting anymore. 

Did we miss an API breaking update perhaps or is there something else going on? We greatly appreciate any help since our community "reads the code" a lot to get an idea of what is going on rather than diving into the docs.

**Comment by lildude** [comment]
Can you provide a link to a repo showing this issue as viewing the [samples](https://github.com/github/linguist/tree/master/samples/Rascal) we ship with Linguist is showing syntax highlighting for me.

**Comment by jurgenvinju** [comment]
Yes. Thanks for looking into this!

* This one does not work: https://github.com/usethesource/rascal/blob/main/src/org/rascalmpl/library/Boolean.rsc
* This sibling file does: https://github.com/usethesource/rascal/blob/main/src/org/rascalmpl/library/DateTime.rsc

**Reply by jurgenvinju** [reply]
I flipped the examples at first. see edit.

**Reply by lildude** [reply]
The issue here is these files without syntax highlighting are being detected as "RouterOS Script" as that language shares the same extension:

```console
➜ github-linguist rascal/src/org/rascalmpl/library/Boolean.rsc
rascal/src/org/rascalmpl/library/Boolean.rsc: 132 lines (96 sloc)
  type:      Text
  mime type: text/plain
  language:  RouterOS Script

```

This is happening because there isn't a heuristic for the `.rsc` extension which means things are falling through to the classifier which is making a guess based on the samples we have for both languages.

You can force the correct language by implementing an [override](https://github.com/github/linguist/blob/master/docs/overrides.md). If you've got the time, we'd be glad to see a heuristic to help with the classification.

This discussion has however highlighted that RouterOS Script's syntax highlighting is no longer working for some reason.

**Reply by jurgenvinju** [reply]
Aha. that makes sense. Thanks. If we can help with a heuristic that's fine. Is there anyone from the RouterOS team that we can contact? Seems like the ideal opportunity to synchronize.

**Reply by lildude** [reply]
@zachstence might be a good start as they added support to Linguist in https://github.com/github/linguist/pull/5945 and I recall there being a challenge with getting the heuristic right so it was dropped.

**Reply by jurgenvinju** [reply]
Hi @zachstence; I was thinking if I make a PR for linguist with a heuristic for Rascal, then you could review it to see if there isn't a colllision with RouterOS script? And I could do the same for your PR with a heuristic for Router OS script? That way we prevent a race to the bottom ;-)

**Reply by jurgenvinju** [reply]
Rascal is a funny language because it is used to implement/analyze/transform/generate other programming languages. So often it contains fragments of other languages in the other language's syntax, or using keywords from the other languages as identifiers in its own. It has, for example, its own embedded dialect of EBNF, but also "concrete syntax" fragments (patterns written in the object language) and multi-line string templates with auto-indent. So writing a heuristic that excludes Rascal is much harder than one that includes it.

Perhaps there are other meta languages in linguist that we could take a hint from. I'll have a look in the list later.

**Reply by zachstence** [reply]
> Hi @zachstence; I was thinking if I make a PR for linguist with a heuristic for Rascal, then you could review it to see if there isn't a colllision with RouterOS script? And I could do the same for your PR with a heuristic for Router OS script? That way we prevent a race to the bottom ;-)

Yeah I'd love to help out here. I'm not super familiar with exactly how the heuristics in linguist work, but I'm sure we could get it worked out together.

**Comment by Cookie-Monster-web** [comment]
hoi jurgen vinju ik ben david vinju hehe
